### PR TITLE
Hw20

### PR DESCRIPTION
**Pull request by Engelko**

### CODE REVIEW

File: Bash_scripts/HW20/group_vars/linux/linux_var.yml [hunk 1]
```diff
@@ -0,0 +1,3 @@
+#Linux vars
+ansible_user: vagrant
```
Comment (mavrik90) [inline Bash_scripts/HW20/group_vars/linux/linux_var.yml:2]: `ansible_user` and `ansible_ssh_private_key_file` can be set up for all hosts, not for each group

File: Bash_scripts/HW20/Vagrantfile [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+  node2.vm.network "public_network", ip: "192.168.1.32"
+ end
+  config.vm.provision "shell" do |s|
+    ssh_pub_key = File.readlines("#{Dir.home}/.ssh/id_rsa.pub").first.strip
+    s.inline = <<-SHELL
+      echo #{ssh_pub_key} >> /home/vagrant/.ssh/authorized_keys
+    SHELL
+  end
+
+end
```
Comment (mavrik90) [inline Bash_scripts/HW20/Vagrantfile]: new line

File: Bash_scripts/HW20/roles/hw201/tasks/centos.yml [hunk 3]
```diff
@@ -0,0 +1,11 @@
+- name: "Install programs on CentOS"
+  ansible.builtin.package:
```
Comment (mavrik90) [inline Bash_scripts/HW20/roles/hw201/tasks/centos.yml]: Is this task `os_family` specific?

File: Bash_scripts/HW20/roles/hw201/tasks/ubuntu.yml [hunk 4]
```diff
@@ -0,0 +1,11 @@
+- name: "Install programs on Ubuntu"
```
Comment (mavrik90) [inline Bash_scripts/HW20/roles/hw201/tasks/ubuntu.yml]: See comment above

File: Bash_scripts/HW20/roles/hw201/tasks/main.yml [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+- name: Install programs on Ubuntu
+  import_tasks: ubuntu.yml
+  
+#Install docker
+- name: Install docker
+  import_tasks: docker.yml 
+   
+#Add user Ubuntu
+- name: Add the user 'hw20' with a sh shell and '/opt/hw20' home directory Ubuntu
+  ansible.builtin.user:
```
Comment (mavrik90) [inline Bash_scripts/HW20/roles/hw201/tasks/main.yml:11]: This is not `os_family` specific. Use variables to make user creating in one task for both OSs

File: Bash_scripts/HW20/roles/hw201/tasks/main.yml [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+   state: present
+  become: true
+  when: ansible_facts['os_family'] == "Debian"
+     
+#Reset connection  
+- name: reset ssh connection to allow user changes to affect 'current login user'
+  meta: reset_connection  
+  
+#Add user CentOS
+- name: Add the user 'hw20' with a sh shell and '/opt/hw20' home directory CentOS
```
Comment (mavrik90) [inline Bash_scripts/HW20/roles/hw201/tasks/main.yml]: See comment above

File: Bash_scripts/HW20/roles/hw202/tasks/main.yml [hunk 7]
```diff
@@ -0,0 +1,51 @@
+---
+#Get info from files
+- name: "Get variables from file"
+  import_tasks: template.yml
+
+#Copy to Ubuntu
+- name: "Copy files to hosts"
```
Comment (mavrik90) [inline Bash_scripts/HW20/roles/hw202/tasks/main.yml:7]: Use variables and make only one task for both OSs

File: Bash_scripts/HW20/group_vars/linux/linux_var.yml (hunk 1 above)
Comment (Engelko) [reply Bash_scripts/HW20/group_vars/linux/linux_var.yml:2]: done. left common variables for all hosts in the Linux group

File: Bash_scripts/HW20/Vagrantfile (hunk 2 above)
Comment (Engelko) [reply Bash_scripts/HW20/Vagrantfile]: if you remove the last end, then vagrant does not work.

syntax error, unexpected end-of-input, expecting end

File: Bash_scripts/HW20/roles/hw201/tasks/centos.yml (hunk 3 above)
Comment (Engelko) [reply Bash_scripts/HW20/roles/hw201/tasks/centos.yml]: fix

File: Bash_scripts/HW20/roles/hw201/tasks/ubuntu.yml (hunk 4 above)
Comment (Engelko) [reply Bash_scripts/HW20/roles/hw201/tasks/ubuntu.yml]: done

File: Bash_scripts/HW20/roles/hw201/tasks/main.yml (hunk 5 above)
Comment (Engelko) [reply Bash_scripts/HW20/roles/hw201/tasks/main.yml:11]: done

File: Bash_scripts/HW20/roles/hw201/tasks/main.yml (hunk 6 above)
Comment (Engelko) [reply Bash_scripts/HW20/roles/hw201/tasks/main.yml]: done

File: Bash_scripts/HW20/roles/hw202/tasks/main.yml (hunk 7 above)
Comment (Engelko) [reply Bash_scripts/HW20/roles/hw202/tasks/main.yml:7]: done

File: Bash_scripts/HW20/group_vars/centos/centos_var.yml [hunk 8]
```diff
@@ -0,0 +1,5 @@
+#Centos vars
+group_hw20: ZONE_2
+group_sudo: wheel,docker
+src_curl: /home/engelko/DevOps_course/Bash_scripts/HW20/roles/hw202/files/curl-7.29.rpm
+dest_curl: /tmp/curl-7.29.rpm
```
Comment (mavrik90) [inline Bash_scripts/HW20/group_vars/centos/centos_var.yml]: New line

File: Bash_scripts/HW20/group_vars/ubuntu/ubuntu_var.yml [hunk 9]
```diff
@@ -0,0 +1,5 @@
+#Ubuntu vars
+group_hw20: ZONE_1
+group_sudo: sudo,docker
+src_curl: /home/engelko/DevOps_course/Bash_scripts/HW20/roles/hw202/files/curl_7.68.deb
+dest_curl: /tmp/curl_7.68.deb
```
Comment (mavrik90) [inline Bash_scripts/HW20/group_vars/ubuntu/ubuntu_var.yml]: new line

File: Bash_scripts/HW20/hosts.yml [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+  linux:
+   hosts:
+    192.168.1.31:
+    192.168.1.32:
+  ubuntu:
+   hosts:
+    192.168.1.31:
+  centos:
+   hosts:
+    192.168.1.32:
```
Comment (mavrik90) [inline Bash_scripts/HW20/hosts.yml]: new line

File: Bash_scripts/HW20/roles/hw201/tasks/main.yml [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+   state: present
+  become: true
+  when: "'linux' in group_names"
+     
+#Reset connection  
+- name: reset ssh connection to allow user changes to affect 'current login user'
+  meta: reset_connection  
+  
+#Pull public git repo
+- name: Pull public git repo as user hw20 to /opt/hw20/
```
Comment (mavrik90) [inline Bash_scripts/HW20/roles/hw201/tasks/main.yml:26]: If you do it via root do not forget to change owner

File: Bash_scripts/HW20/group_vars/ubuntu/ubuntu_var.yml (hunk 9 above)
Comment (mavrik90) [inline Bash_scripts/HW20/group_vars/ubuntu/ubuntu_var.yml]: var name have no sense

File: Bash_scripts/HW20/roles/hw202/tasks/main.yml [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+#Get info from files
+- name: "Get variables from file"
+  import_tasks: template.yml
+
+#Copy to group linux
+- name: "Copy files to hosts"
+  become: true
+  become_user: root
+  ansible.builtin.copy:
+   src: "{{ src_curl }}"
```
Comment (mavrik90) [inline Bash_scripts/HW20/roles/hw202/tasks/main.yml]: Current task implemented for single file copy. Improve for multiple file copy. Use `loop`

File: Bash_scripts/HW20/roles/hw202/tasks/main.yml [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+ 
+  #Install Curl on CentOS
+- name: "Install a .rpm package"
+  become: true
+  ansible.builtin.yum:
+    name: "{{ dest_curl }}"
+    state: present
+    update_cache: yes
+    allow_downgrade: yes
+  when: ansible_facts['os_family'] == "RedHat"
```
Comment (mavrik90) [inline Bash_scripts/HW20/roles/hw202/tasks/main.yml]: Please do this https://stackoverflow.com/questions/44704968/visual-studio-code-insert-new-line-at-the-end-of-files :) Save your and my time ;)

File: Bash_scripts/HW20/roles/hw202/tasks/template.yml [hunk 14]
```diff
@@ -0,0 +1,7 @@
+- name: "Template a file to /tmp/HW20_FILE_INFO"
+  ansible.builtin.template:
+    src: /home/engelko/DevOps_course/Bash_scripts/HW20/roles/hw202/templates/main.j2
```
Comment (mavrik90) [inline Bash_scripts/HW20/roles/hw202/tasks/template.yml]: Do not use full path

File: Bash_scripts/HW21/in_hw20/group_vars/centos/centos_var.yml [hunk 15]
```diff
@@ -0,0 +1,4 @@
+#Centos vars
+ansible_user: vagrant
```
Comment (mavrik90) [inline Bash_scripts/HW21/in_hw20/group_vars/centos/centos_var.yml]: Do you need this var here if you already have it under `linux` group?

File: Bash_scripts/HW21/in_hw20/group_vars/ubuntu/ubuntu_var.yml [hunk 16]
```diff
@@ -0,0 +1,4 @@
+#Ubuntu vars
+ansible_user: vagrant
+ansible_ssh_private_key_file: /home/engelko/.ssh/id_rsa
```
Comment (mavrik90) [inline Bash_scripts/HW21/in_hw20/group_vars/ubuntu/ubuntu_var.yml]: Do you need this var here if you already have it under linux group?

File: Bash_scripts/HW21/in_hw20/roles/hw201/tasks/centos.yml [hunk 17]
```diff
@@ -0,0 +1,11 @@
+- name: "Install programs on CentOS"
```
Comment (mavrik90) [inline Bash_scripts/HW21/in_hw20/roles/hw201/tasks/centos.yml]: Forget to delete?

Comment (Engelko) [reply Bash_scripts/HW21/in_hw20/roles/hw201/tasks/centos.yml]: Sorry, I forgot to delete, everything is fixed.

File: Bash_scripts/HW21/in_hw20/group_vars/centos/centos_var.yml (hunk 15 above)
Comment (Engelko) [reply Bash_scripts/HW21/in_hw20/group_vars/centos/centos_var.yml]: Sorry, this is not the job, this is hw21, everything is fixed, the directory is deleted

File: Bash_scripts/HW21/in_hw20/group_vars/ubuntu/ubuntu_var.yml (hunk 16 above)
Comment (Engelko) [reply Bash_scripts/HW21/in_hw20/group_vars/ubuntu/ubuntu_var.yml]: Sorry, this is not the job, this is hw21, everything is fixed, the directory is deleted

File: Bash_scripts/HW20/roles/hw202/tasks/main.yml (hunk 13 above)
Comment (Engelko) [reply Bash_scripts/HW20/roles/hw202/tasks/main.yml]: thanks, installed

File: Bash_scripts/HW20/group_vars/centos/centos_var.yml (hunk 8 above)
Comment (Engelko) [reply Bash_scripts/HW20/group_vars/centos/centos_var.yml]: fixed

File: Bash_scripts/HW20/group_vars/ubuntu/ubuntu_var.yml (hunk 9 above)
Comment (Engelko) [reply Bash_scripts/HW20/group_vars/ubuntu/ubuntu_var.yml]: fixed

File: Bash_scripts/HW20/hosts.yml (hunk 10 above)
Comment (Engelko) [reply Bash_scripts/HW20/hosts.yml]: fixed

File: Bash_scripts/HW20/roles/hw201/tasks/main.yml (hunk 11 above)
Comment (Engelko) [reply Bash_scripts/HW20/roles/hw201/tasks/main.yml:26]: you don't need to use root here at all. deleted.

File: Bash_scripts/HW20/roles/hw202/tasks/main.yml (hunk 12 above)
Comment (Engelko) [reply Bash_scripts/HW20/roles/hw202/tasks/main.yml]: done. I add loop for copy task

File: Bash_scripts/HW20/roles/hw202/tasks/template.yml (hunk 14 above)
Comment (Engelko) [reply Bash_scripts/HW20/roles/hw202/tasks/template.yml]: done

File: Bash_scripts/HW20/group_vars/ubuntu/ubuntu_var.yml (hunk 9 above)
Comment (mavrik90) [reply Bash_scripts/HW20/group_vars/ubuntu/ubuntu_var.yml]: Not really

File: Bash_scripts/HW20/hosts.yml (hunk 10 above)
Comment (mavrik90) [reply Bash_scripts/HW20/hosts.yml]: nope

File: Bash_scripts/HW20/roles/hw201/tasks/docker.yml [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+      ansible.builtin.apt_key:
+       url: https://download.docker.com/linux/ubuntu/gpg
+       state: present
+       
+    - name: "Add repos"
+      ansible.builtin.apt_repository:
+       repo: deb https://download.docker.com/linux/ubuntu focal stable
+       state: present
+       
+    - name: Install and Start docker engine
```
Comment (mavrik90) [inline Bash_scripts/HW20/roles/hw201/tasks/docker.yml]: I think it make sense to use handler here to start docker. Docker install should be separated

File: Bash_scripts/HW20/roles/hw201/tasks/docker.yml [hunk 19]
```diff
@@ -0,0 +1,43 @@
...
+    - name: "Add Docker repo"
+      ansible.builtin.get_url:
+       url: https://download.docker.com/linux/centos/docker-ce.repo
+       dest: /etc/yum.repos.d/docer-ce.repo
+      
+    - name: Install and Start docker engine
+      import_tasks: docker_start.yml
+       
+  when: ansible_facts['os_family'] == "RedHat"
+  become: true
```
Comment (mavrik90) [inline Bash_scripts/HW20/roles/hw201/tasks/docker.yml]: new line

File: Bash_scripts/HW20/roles/hw202/tasks/main.yml [hunk 20]
```diff
@@ -0,0 +1,38 @@
...
+
```
Comment (mavrik90) [inline Bash_scripts/HW20/roles/hw202/tasks/main.yml]: fix end of file

File: Bash_scripts/HW20/roles/hw202/tasks/main.yml [hunk 21]
```diff
@@ -0,0 +1,38 @@
...
+- name: "Get variables from file"
+  import_tasks: template.yml
+
+#Copy to group linux
+- name: "Copy files to hosts"
+  become: true
+  become_user: root
+  ansible.builtin.copy:
+   src: "{{ item }}"
+   dest: "{{ dest_curl }}"
```
Comment (mavrik90) [inline Bash_scripts/HW20/roles/hw202/tasks/main.yml]: What if files have different dests?

File: Bash_scripts/HW20/roles/hw202/tasks/main.yml [hunk 22]
```diff
@@ -0,0 +1,38 @@
...
+   owner: hw20
+   mode: 0744
+  loop: 
+   - "{{ src_curl }}"
+  when: "'linux' in group_names"
```
Comment (mavrik90) [inline Bash_scripts/HW20/roles/hw202/tasks/main.yml]: Do you need `when` is this case?

File: Bash_scripts/HW20/Vagrantfile (hunk 2 above)
Comment (mavrik90) [reply Bash_scripts/HW20/Vagrantfile]: looks like this is not truу
Try to play with it.

File: Bash_scripts/HW20/roles/hw201/tasks/docker_start.yml [hunk 23]
```diff
@@ -0,0 +1,16 @@
+#Install Docker engine and Start docker service
+
+- name: "Install Docker Engine"
```
Comment (mavrik90) [inline Bash_scripts/HW20/roles/hw201/tasks/docker_start.yml]: This should be under docker.yml as it's not related to start action

File: Bash_scripts/HW20/roles/hw201/tasks/docker_start.yml [hunk 24]
```diff
@@ -0,0 +1,16 @@
...
+  ansible.builtin.package:
+    name: "{{item}}"
+    update_cache: yes
+  loop:
+    - docker-ce
+    - docker-ce-cli
+    - containerd.io
+             
+- name: "Add Docker to start up"
```
Comment (mavrik90) [inline Bash_scripts/HW20/roles/hw201/tasks/docker_start.yml]: This should be done as handler

**Comment by mavrik90** [comment]
Docker service should be enabled for startup

File: Bash_scripts/HW20/roles/hw201/tasks/docker.yml [hunk 25]
```diff
@@ -0,0 +1,43 @@
...
+       state: present
+       
+    - name: Install and Start docker engine
+      import_tasks: docker_start.yml
+           
+  when: ansible_facts['os_family'] == "Debian"
+  become: true
+
+#CentOS
+- name: "Install Docker on Ubuntu"
```
Comment (mavrik90) [inline Bash_scripts/HW20/roles/hw201/tasks/docker.yml]: Incorrect name

File: Bash_scripts/HW20/group_vars/ubuntu/ubuntu_var.yml (hunk 9 above)
Comment (Engelko) [reply Bash_scripts/HW20/group_vars/ubuntu/ubuntu_var.yml]: now fixed it for sure

File: Bash_scripts/HW20/hosts.yml (hunk 10 above)
Comment (Engelko) [reply Bash_scripts/HW20/hosts.yml]: now fixed it for sure

File: Bash_scripts/HW20/group_vars/ubuntu/ubuntu_var.yml (hunk 9 above)
Comment (Engelko) [reply Bash_scripts/HW20/group_vars/ubuntu/ubuntu_var.yml]: the full path meant? I don't quite understand what you mean

File: Bash_scripts/HW20/roles/hw201/tasks/docker.yml (hunk 19 above)
Comment (Engelko) [reply Bash_scripts/HW20/roles/hw201/tasks/docker.yml]: complete

File: Bash_scripts/HW20/roles/hw201/tasks/docker.yml (hunk 25 above)
Comment (Engelko) [reply Bash_scripts/HW20/roles/hw201/tasks/docker.yml]: changed

File: Bash_scripts/HW20/roles/hw201/tasks/docker.yml (hunk 18 above)
Comment (Engelko) [reply Bash_scripts/HW20/roles/hw201/tasks/docker.yml]: added handler to start docker

File: Bash_scripts/HW20/roles/hw202/tasks/main.yml (hunk 21 above)
Comment (Engelko) [reply Bash_scripts/HW20/roles/hw202/tasks/main.yml]: they have different paths. Add variables to different hotspot groups. Each host group will have its own dest.path

File: Bash_scripts/HW20/roles/hw202/tasks/main.yml (hunk 22 above)
Comment (Engelko) [reply Bash_scripts/HW20/roles/hw202/tasks/main.yml]: in this example, rather no. chenged.

File: Bash_scripts/HW20/roles/hw201/tasks/docker_start.yml (hunk 23 above)
Comment (Engelko) [reply Bash_scripts/HW20/roles/hw201/tasks/docker_start.yml]: Made hendler to start, returned the rest of the installation to docker.yml

File: Bash_scripts/HW20/roles/hw201/tasks/docker_start.yml (hunk 24 above)
Comment (Engelko) [reply Bash_scripts/HW20/roles/hw201/tasks/docker_start.yml]: Changed

File: Bash_scripts/HW20/roles/hw201/tasks/docker.yml [hunk 26]
```diff
@@ -0,0 +1,44 @@
...
+       state: present
+      with_items:
+       - "{{ url_key}}"
+
+    - name: "Add repos"
+      ansible.builtin.apt_repository:
+       repo: "{{ repo }}"
+       state: present        
+  when: ansible_facts['os_family'] == "Debian"
+  become: true
```
Comment (mavrik90) [inline Bash_scripts/HW20/roles/hw201/tasks/docker.yml:23]: become can go to `import_task` level

**Review by mavrik90 — APPROVED** [review]
Well done!!!

**Pull request merged**